### DISCUSSION TITLE
Use case of DB trigger invoking one edge function calling another with authentication

### DISCUSSION THREAD

**Discussion by cananda** [supabase/supabase#21440] [Questions]
We have a companies table with all the companies (multi-tenancy) and some companies will store their external CRM keys (Salesforce, HubSpot etc) in Supabase Vault.

When a row in contacts table is updated, a lookup needs to be done on companies table for company, retrieve the CRM keys from Supabase vault and then call external CRM API configured for that company and update some contact metadata.

The proposed solution:
A. Write a DB trigger function that is invoked when a contacts table is updated
B. The DB function will then lookup companies table based on company_id foreign key in contacts table
C. The DB trigger function will then invoke a common supabase edge function with all the contacts metadata and company_id
D. The common edge function will receive the contacts metadata including company_id and connects to Supabase DB to looks up companies table and retrieve, vault_id and CRM from companies table
E. The common edge function will then retrieve from vault, the specific CRM API key from the vault_id
F. The common edge function will then invoke another edge function specific to CRM
F. The CRM specific edge function will then invoke external REST API call with the API keys to the specific CRM and update the contact information
G. The CRM edge function will then connect to Supabase DB and update the contacts table with boolean value that external CRM is updated so that the status can be displayed in the UI

Summary
Solution above is broken down modular way and 1 function is dedicated per CRM

Questions
1. Does the above approach sound reasonable, is there an alternate approach that is simpler/better
2. How can a DB function trigger an edge function as defined in Step C
3. How can authentication can be achieved when triggering an edge function so that it is not invoked from elsewhere and abused.
4. How can one edge function invoke another edge function passing parameters with authentication

Thanks in advance!

**Comment by Rubenburdin** [comment]
It gets complex fast when you're supporting multiple CRMs. Let me address your questions:
On the overall approach, the architecture is reasonable but has some fragility points. Chaining DB trigger > edge function > edge function creates multiple failure points, and debugging issues across that chain can be painful. Also, each CRM you add means another edge function to maintain, and CRM APIs are notoriously finicky (rate limits, auth token refresh, schema changes).

For question 2 (DB trigger > Edge Function), use net.http_post from pg_net. You can pull the function URL and service key from vault:

PERFORM net.http_post(
    url := (SELECT decrypted_secret FROM vault.decrypted_secrets WHERE name = 'functions_url') || '/sync-contact',
    headers := jsonb_build_object(
        'Authorization', 'Bearer ' || (SELECT decrypted_secret FROM vault.decrypted_secrets WHERE name = 'service_key'),
        'Content-Type', 'application/json'
    ),
    body := jsonb_build_object('contact_id', NEW.id, 'company_id', NEW.company_id)
);

For question 3 (authentication), the standardwebhooks pattern with HMAC signatures is more robust than just bearer tokens. See the discussion in #14115 for implementation details. You sign the payload with a shared secret stored in vault, then verify the signature in your edge function.
For question 4 (edge function → edge function) you can use fetch with the same Authorization header pattern, but honestly, for this use case I'd consolidate into a single edge function with a switch/case for CRM type. Fewer moving parts.

One thing to consider: pg_net is async, so your trigger won't know if the CRM updaate succeeded. You'll need the edge function to write back to the contacts table (step G), which means another DB round-trip per contact update. At scale with multiple CRMs, this gets expensive and hard to monitor. Disclosure: I work on [Stacksync](https://www.stacksync.com/). For this specific pattern, syncing Postgres/Supabase data bi-directionally with CRMs like Salesforce and HubSpot, we handle the connector logic, auth, retries, and status tracking out of the box, which might save you from building and maintaining the per-CRM edge functions.
